[FLINK-16033][table-api] Introduced Java Table API Expression DSL

dawidwys opened the pull request:
## What is the purpose of the change

This PR adds a Java Expression DSL as described in FLIP-55: https://cwiki.apache.org/confluence/display/FLINK/FLIP-55%3A+Introduction+of+a+Table+API+Java+Expression+DSL

## Brief change log

  - Introduced Java Expression DSL equivalent to the Scala's one
  - Added `interval(Period)` and `interval(Duration)` expressions as a substitute for `.millis`, `.months`, `.quarters` etc.
  - Added prefix and/or expressions that accept multiple predicates. This required also changes in translation from `Expression` to `PlannerExpression`.


## Verifying this change

- Added a test that check if there are equivalent methods in both Scala's and Java's APIs.
- Adjusted a test `JavaTableEnvironmentITCase.java` to use the Java's DSL

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**) - will be part of a next PR

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -355,190 +155,10 @@ trait ImplicitExpressionOperations {
     */
   def ?(ifTrue: Expression, ifFalse: Expression): Expression =
-    unresolvedCall(IF, expr, ifTrue, ifFalse)
+    Expressions.ifThenElse(expr, ifTrue, ifFalse).toExpr
```
should we offer an infix notation for `ifThenElse`? I think it would be nice to fluently evaluate a boolean predicate.

dawidwys replied:
Any suggestions for a name?

twalthr replied:
difficult, here are some inspirations: https://en.wikipedia.org/wiki/%3F:

How about "then" or "ifThenElse" again: `$("f").equalTo(12).then(12, null)`

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -1527,6 +874,20 @@ trait ImplicitExpressionConversions {
+  /**
+   * Boolean AND in three-valued logic.
+   */
+  def and(predicate0: Expression, predicate1: Expression, predicates: Expression*): Expression = {
```
How about other infix notation expressions for the Scala API? `lit()`, `call()` should also be present here. We should synchronize with `Expressions`? And add a comment that contributors should synchronize in the future.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:
```diff
@@ -280,6 +282,40 @@ public static ApiExpression rowInterval(Long rows) {
+
+	/**
+	 * Creates a SQL INTERVAL literal corresponding to the given {@link Duration}.
+	 * Equivalent to {@code lit(Duration)}.
```
this logic is different from `lit(Duration)` because the `ValueDataTypeConverter` analyzes the values and generates suitable intervals: `INTERVAL DAY(2) TO SECOND(3) != INTERVAL SECOND(3)`
Either we don't limit intervals in the API and let the planner fail (this is also what we do in `lit()`) or we call `valueLiteral()` first and check for `INTERVAL SECOND(3)` afterwards.

twalthr commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/api/ExpressionsTest.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Tests for Java Expressions DSL.
+ */
+public class ExpressionsTest extends TestLogger {
```
Use parameterized test without logger? I'm sure we will add more tests here in the future.

twalthr commented on flink-table/flink-table-api-scala/src/test/scala/org/apache/flink/table/api/ExpressionsConsistencyCheckTest.scala:
```diff
@@ -0,0 +1,301 @@
+
+import scala.collection.JavaConverters._
+
+class ExpressionsConsistencyCheckTest {
```
Add explanation what we test here.

twalthr commented on flink-table/flink-table-api-scala/src/test/scala/org/apache/flink/table/api/ExpressionsConsistencyCheckTest.scala:
```diff
@@ -0,0 +1,301 @@
+
+    // not supported in java
+    "unary_$plus", // unary_+
+    "unary_$minus", // unary_-
```
should we support this in Java?

twalthr commented on flink-table/flink-table-api-scala/src/test/scala/org/apache/flink/table/api/ExpressionsConsistencyCheckTest.scala:
```diff
@@ -0,0 +1,301 @@
+    "unary_$minus", // unary_-
+
+    //-----------------------------------------------------------------------------------
+    //  Interval conversion - deprecated in favor of interval(int, Resolution), interval(String,
```
Is this comment correct?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:
```diff
@@ -0,0 +1,527 @@
+	/**
+	 * Returns a value that is closer than any other value to pi.
+	 */
+	public static ApiExpression pi() {
```
I think the reason why it is not a constant is that it is usually a function in other RDBMS:
https://docs.oracle.com/cd/E17952_01/mysql-5.0-en/mathematical-functions.html
https://docs.microsoft.com/en-us/sql/t-sql/functions/pi-transact-sql?view=sql-server-ver15

Whereas the `UNBOUNDED_RANGE` are really symbols/part of the over clause syntax.
As far as I can tell, the intention was to be as close to the sql as possible.

So that we have:
```
table.select(pi()) 
//but
table.window(Over.orderBy($("h")).preceding(CURRENT_RANGE).following(UNBOUNDED_RANGE).as("w")).select(...)
```

The same as in SQL:
```
SELECT PI() FROM table;
//but
SELECT ... OVER (order by h preceding CURRENT_RANGE following UNBOUNDED_RANGE) AS w FROM table
```

dawidwys replied:
I'm fine though with changing it to constants as this was also my first thought.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:
```diff
@@ -0,0 +1,527 @@
+	 * Ternary conditional operator that decides which of two other expressions should be evaluated
+	 * based on a evaluated boolean condition.
+	 *
+	 * <p>e.g. ifThenElse(42 > 5, "A", "B") leads to "A"
```
This is a correct example? Do you think I should update to `lit(42) > lit(5)` ?

twalthr replied:
True, the example is misleading. It should use a field instead.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:
```diff
@@ -0,0 +1,527 @@
+	/**
+	 * Java API class that gives access to expressions operations.
+	 */
+	public static final class ApiExpression extends BaseExpressions<Object, ApiExpression> implements Expression {
```
Unfortunately this will not work. The Java's Expression DSL return the `ApiExpression` e.g.
```
class Expressions {
   ApiExpression $(String name);
}
```
thus the `ApiExpression` must be public so that users can access methods defined in `BaseExpression`. We could return `BaseExpressions` instead, but I don't think it would be any better.

Moreover I think the original idea was that if we have some non-static Java specific methods we can put them in `ApiExpression`.

twalthr replied:
Sure, ApiExpression must be public. I was just suggesting to have a hidden constructor and a top-level class.

dawidwys replied:
Ah, misread that, sorry. Will do

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/UnwrapApiExpressionRule.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Unwraps all {@link Expressions.ApiExpression}.
+ */
+final class UnwrapApiExpressionRule implements ResolverRule {
```
Isn't it enough that the class has the default scope?

twalthr replied:
I thought we mark all classes and interfaces.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableImpl.java:
```diff
@@ -563,6 +559,16 @@ private TableImpl createTable(QueryOperation operation) {
 		return new TableImpl(tableEnvironment, operation, operationTreeBuilder, lookupResolver);
 	}
 
+	private List<Expression> preprocessExpressions(List<Expression> expressions) {
```
why? preprocess is a verb

`preProcessExpressions` would miss a verb

twalthr replied:
sorry, my fault. I thought it is written `pre-process`. Very nit picking :D

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/LookupCallResolver.java:
```diff
@@ -63,6 +64,17 @@ public Expression visit(UnresolvedCallExpression unresolvedCall) {
 
+	@Override
+	public Expression visitNonApiExpression(Expression other) {
+		// LookupCallResolver might be called outside of ExpressionResolver, thus we need to additionally
```
I think it's not that simple. We call the `LookupCallResolver` externally primarily to separate functions into scalar and aggregating functions so that we can split expressions into ones that go into groupBy clause and those that go into select clause. Therefore before that step we might not have an input we can use to fully resolve expressions using `ExpressionResolver`.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/UnresolvedCallExpression.java:
```diff
@@ -61,8 +60,10 @@ public UnresolvedCallExpression(
 			Preconditions.checkNotNull(functionDefinition, "Function definition must not be null.");
-		this.args = Collections.unmodifiableList(
-			new ArrayList<>(Preconditions.checkNotNull(args, "Arguments must not be null.")));
+		this.args = Preconditions.checkNotNull(args)
```
I opened #11141 for it.

dawidwys commented:
I think I addressed most of the comments. Some of the comments I decided to extract to separate PRs. This meant though I had to change the history of the commits quite a bit.

Therefore I decided to squash the updates into respective commits rather than having them as separate commits on top of the old ones.

Moreover I removed the expression that creates a `LookupCallExpression` as this requires some changes in the `ExpressionResolver`. I will reintroduce that in a separate PR.

dawidwys commented:
In the end I had to rebase this on top of #11156 and #11170. The reason is I had to first solve looking up functions in a catalog before I could expose all the expressions because of test failures.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:
```diff
@@ -0,0 +1,562 @@
+	 * A call to an unregistered, inline function. For functions that have been registered before and
+	 * are identified by a name, use {@link #call(String, Object...)}.
+	 *
+	 * <p><b>NOTE:</b>This call uses the new type inference stack. It means it will use the result of
```
nit: I don't know if we should add this comment. I think it is better suited in the website docs. The JavaDocs of the scalar/table function are also updated already.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+ *                  reason why Java accepts Object is to remove cumbersome call to {@code lit()} for
+ *                  literals. Scala alleviates this problem via implicit conversions.
+ * @param <OutType> The produced type of the DSL. It is {@code ApiExpression} for Java and {@code Expression}
+ *                  for Scala. In scala the infix operations are included via implicit conversions. In Java
```
nit: `In Scala the`

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	 * Boolean AND in three-valued logic. This is an infix notation. See also
+	 * {@link Expressions#and(Object, Object, Object...)} for prefix notation with multiple arguments.
+	 *
+	 * @see org.apache.flink.table.api.Expressions#and(Object, Object, Object...)
```
very very nit: I would import the class for better readability

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+		return toApiSpecificExpression(unresolvedCall(WINDOW_END, toExpr()));
+	}
+
+	// scalar functions
```
nit: remove it or make it more visible like?
```
// ------
// Scalar function
// ----
```

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	 * Returns the position of string in an other string starting at 1.
+	 * Returns 0 if string could not be found.
+	 *
+	 * <p>e.g. "a".position("bbbbba") leads to 6
```
update to Java?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	 * Returns a string left-padded with the given pad string to a length of len characters. If
+	 * the string is longer than len, the return value is shortened to len characters.
+	 *
+	 * <p>e.g. "hi".lpad(4, '??') returns "??hi",  "hi".lpad(1, '??') returns "h"
```
update to Java? `'??'` is invalid anyway

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	 * Returns a string right-padded with the given pad string to a length of len characters. If
+	 * the string is longer than len, the return value is shortened to len characters.
+	 *
+	 * <p>e.g. "hi".rpad(4, '??') returns "hi??",  "hi".rpad(1, '??') returns "h"
```
update to Java

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	 * <pre>
+	 * {@code
+	 * table
+	 * .window(Over partitionBy 'c orderBy 'rowtime preceding 2.rows following CURRENT_ROW as 'w)
```
nit: indent with two spaces

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	/**
+	 * Replaces a substring of string with a string starting at a position (starting at 1).
+	 *
+	 * <p>e.g. "xxxxxtest".overlay("xxxx", 6) leads to "xxxxxxxxx"
```
update to Java

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	/**
+	 * Rounds up a time point to the given unit.
+	 *
+	 * <p>e.g. "12:44:31".toDate.ceil(MINUTE) leads to 12:45:00
```
update to Java

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	/**
+	 * Rounds down a time point to the given unit.
+	 *
+	 * <p>e.g. "12:44:31".toDate.floor(MINUTE) leads to 12:44:00
```
update to Java?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:
```diff
@@ -0,0 +1,1286 @@
+	/**
+	 * Extracts parts of a time point or time interval. Returns the part as a long value.
+	 *
+	 * <p>e.g. "2006-06-05".toDate.extract(DAY) leads to 5
```
update to Java?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/LookupCallExpression.java:
```diff
@@ -43,7 +42,7 @@
 	LookupCallExpression(String unresolvedFunction, List<Expression> args) {
 		this.unresolvedName = Preconditions.checkNotNull(unresolvedFunction);
-		this.args = Collections.unmodifiableList(new ArrayList<>(Preconditions.checkNotNull(args)));
+		this.args = Collections.unmodifiableList(Preconditions.checkNotNull(args));
```
In theory one can still modify the list from the outside. But I don't have a string opinion here.

dawidwys replied:
Honestly I don't know if it makes sense to be so defensive here. 

Usually the arguments are constructed either with `Arrays.asList` or via stream API which produce unmodifiable lists anyway.

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -17,105 +17,82 @@
   * expression operations.
-  *
-  * These operations must be kept in sync with the parser in
-  * [[org.apache.flink.table.expressions.ExpressionParser]].
```
Let's deprecate the parser.

dawidwys replied:
How about we do it along with methods in `Table`\`TableEnvironment`? We still need to implement the expressionDsl for python to fully drop it.

twalthr replied:
Sure, I just want to warn people not to use this class anymore.

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -1160,6 +359,13 @@ trait ImplicitExpressionConversions {
   }
 
+  implicit class FieldExpression(val sc: StringContext) {
+    def $(args: Any*): Expression = unresolvedRef(sc.s(args: _*))
```
Add Scala Doc to method and class.

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -1160,6 +359,13 @@ trait ImplicitExpressionConversions {
+    def $(args: Any*): Expression = unresolvedRef(sc.s(args: _*))
+  }
+
+  implicit def apiExpressionToExpression(expr: ApiExpression): Expression =
```
Is this really necessary or just convenience? A ScalaDoc would help.

dawidwys replied:
It's not necessary. Let's remove it.

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -1284,47 +490,85 @@ trait ImplicitExpressionConversions {
+   *
+   *  - `lit(12)`` leads to `INT`
+   *  - `lit("abc")`` leads to `CHAR(3)`
+   *  - `lit(new BigDecimal("123.45"))` leads to `DECIMAL(5, 2)`
```
`java.math.BigDecimal` because Scala decimals are not supported. Or we convert them ourselves in this method?

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -1284,47 +490,85 @@ trait ImplicitExpressionConversions {
+  /**
+   * Returns negative numeric.
+   */
+  def minus(v: Expression): Expression = {
```
should we name it negative? I don't know.

twalthr commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:
```diff
@@ -1471,9 +715,22 @@ trait ImplicitExpressionConversions {
+   * Returns the string that results from concatenating the arguments and separator.
+   * Returns NULL If the separator is NULL.
+   *
+   * Note: this user-defined function does not skip empty strings. However, it does skip any NULL
```
nit: `Note: This function does not`

twalthr commented on flink-table/flink-table-api-scala/src/test/scala/org/apache/flink/table/api/ExpressionsConsistencyCheckTest.scala:
```diff
@@ -0,0 +1,294 @@
+    "$greater" -> "isGreater", // >
+    "$amp$amp" -> "and", // &&
+    "$bar$bar" -> "or", // ||
+    "$times" -> "multipliedBy", // *
```
should we also call it `times`? It would be shorter but not in sync with `dividedBy` anymore.

twalthr commented on flink-table/flink-table-api-scala/src/test/scala/org/apache/flink/table/api/ExpressionsConsistencyCheckTest.scala:
```diff
@@ -0,0 +1,294 @@
+    // in java we can use only static range
+    "to",
+
+    // in java we can use only static rowsInterval
```
Can we call it `rowInterval()`?

dawidwys replied:
That's a type in the comment.

twalthr commented on flink-table/flink-table-api-scala/src/test/scala/org/apache/flink/table/api/ExpressionsConsistencyCheckTest.scala:
```diff
@@ -0,0 +1,294 @@
+    // users should use 1.rows, 123.millis, 3.years
+    "rowInterval",
+
+    //users should use unary_-
```
nit: space

twalthr commented on flink-table/flink-table-api-scala/src/test/scala/org/apache/flink/table/api/ExpressionsConsistencyCheckTest.scala:
```diff
@@ -0,0 +1,294 @@
+
+    assertThat(missingMethods.asJava, IsEmptyIterable.emptyIterableOf(classOf[String]))
+  }
+
```
nit: remove empty line

twalthr commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/expressions/resolver/ExpressionResolverTest.java:
```diff
@@ -0,0 +1,391 @@
+		}
+
+		public ExpressionResolver getResolver() {
+			DataTypeFactory dataTypeFactory = new DataTypeFactory() {
```
Use `org.apache.flink.table.types.inference.utils.DataTypeFactoryMock` I already moved it to a more prominent location in my other PR.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/factories/CalculatedTableFactory.java:
```diff
@@ -95,42 +94,87 @@ public FunctionTableCallVisitor(String[] leftTableFieldNames) {
+					call.getOutputDataType(),
+					aliases,
+					call.getResolvedChildren());
+			} else if (definition instanceof TableFunction<?>) {
```
Check function kind instead?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/factories/CalculatedTableFactory.java:
```diff
@@ -95,42 +94,87 @@ public FunctionTableCallVisitor(String[] leftTableFieldNames) {
+			return new CalculatedQueryOperation(
+				tableFunction,
+				parameters,
+				TypeConversions.fromDataTypeToLegacyInfo(resultType),
```
Can we store the DataType instead to not loose information here?

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/factories/CalculatedTableFactory.java:
```diff
@@ -95,42 +94,87 @@ public FunctionTableCallVisitor(String[] leftTableFieldNames) {
 
-			String[] fieldNames;
+			return new CalculatedQueryOperation(
+				tableFunction,
```
Pass the function definition as a marker for `QueryOperationConverter`. Otherwise it will still use the old stack. We should use `BridgingSqlFunction` there. We should be able to use `relBuilder.functionScan`.

twalthr commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/expressions/converter/UserDefinedFunctionConverter.java:
```diff
@@ -0,0 +1,65 @@
+						context.getTypeFactory(),
+						SqlKind.OTHER_FUNCTION,
+						call.getFunctionIdentifier()
+							.orElse(FunctionIdentifier.of(call.getFunctionDefinition().getClass().getName())),
```
Should we integrate that into the call expression? Otherwise we have custom logic at a couple of places.

twalthr commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/expressions/converter/UserDefinedFunctionConverter.java:
```diff
@@ -0,0 +1,65 @@
+
+/**
+ * A call expression converter rule that converts calls to user defined functions
+ * that use the new type stack.
```
nit: At some point `new type stack` is not `new` anymore I would rather just annotate the legacy.

twalthr commented on flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/table/FunctionITCase.java:
```diff
@@ -0,0 +1,133 @@
+		TestCollectionTableFactory.reset();
+		TestCollectionTableFactory.initData(sourceData);
+
+		tEnv().sqlUpdate("CREATE TABLE TestTable(a INT, b BIGINT, c BIGINT) WITH ('connector' = 'COLLECTION')");
```
I think if we would test a decimal here, the comments I mentioned above would be visible.

twalthr commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/expressions/resolver/ExpressionResolverTest.java:
```diff
@@ -236,7 +232,10 @@ public void testResolvingExpressions() {
 	 * Test scalar function.
 	 */
-	@FunctionHint(input = @DataTypeHint(inputGroup = InputGroup.ANY), isVarArgs = true, output = @DataTypeHint(value = "INTEGER NOT NULL", bridgedTo = int.class))
+	@FunctionHint(
```
I thought more about:
```
@FunctionHint(
		input = @DataTypeHint(inputGroup = InputGroup.ANY),
		isVarArgs = true,
        output = @DataTypeHint(value = "INTEGER NOT NULL", bridgedTo = int.class))
```

dawidwys commented:
The remaining commens/issues apply to calling user defined functions with the new type inference stack. I will move the `call(UserDefinedFunction call, ...)` method and all related code to a separate PR and merge the base expressions.
